Userland/Libraries/LibWeb/XHR/XMLHttpRequest.cpp

```diff
@@ -311,19 +312,41 @@ Optional<MimeSniff::MimeType> XMLHttpRequest::extract_mime_type(Fetch::HeaderLis
 }
 
 // https://fetch.spec.whatwg.org/#concept-bodyinit-extract
-static XMLHttpRequest::BodyWithType extract_body(XMLHttpRequestBodyInit const& body)
+static Fetch::BodyWithType extract_body(XMLHttpRequestBodyInit const& body_init)
 {
-    XMLHttpRequest::BodyWithType body_with_type {};
-    body.visit(
+    // FIXME: 1. Let stream be object if object is a ReadableStream object. Otherwise, let stream be a new ReadableStream, and set up stream.
+    Fetch::Body::ReadableStreamDummy stream {};
+    // FIXME: 2. Let action be null.
+    // 3. Let source be null.
+    Fetch::Body::SourceType source {};
+    // 4. Let length be null.
+    Optional<u64> length {};
+    // 5. Let type be null.
+    Optional<ByteBuffer> type {};
+
+    // 6. Switch on object.
+    body_init.visit(
         [&](NonnullRefPtr<URL::URLSearchParams> const& url_search_params) {
-            body_with_type.body = url_search_params->to_string().to_byte_buffer();
-            body_with_type.type = "application/x-www-form-urlencoded;charset=UTF-8";
+            // Set source to the result of running the application/x-www-form-urlencoded serializer with object’s list.
+            source = url_search_params->to_string().to_byte_buffer();
+            // Set type to `application/x-www-form-urlencoded;charset=UTF-8`.
+            type = String { "application/x-www-form-urlencoded;charset=UTF-8" }.to_byte_buffer();
         },
-        [&](String const& string) {
-            body_with_type.body = string.to_byte_buffer();
-            body_with_type.type = "text/plain;charset=UTF-8";
+        [&](String const& scalar_value_string) {
+            // NOTE: AK::String is always UTF-8.
+            // Set source to the UTF-8 encoding of object.
+            source = scalar_value_string.to_byte_buffer();
+            // Set type to `text/plain;charset=UTF-8`.
+            type = String { "text/plain;charset=UTF-8" }.to_byte_buffer();
         });
-    return body_with_type;
+
+    // FIXME: 7. If source is a byte sequence, then set action to a step that returns source and length to source’s length.
+    // FIXME: 8. If action is non-null, then run these steps in in parallel:
+
+    // 9. Let body be a body whose stream is stream, source is source, and length is length.
+    auto body = Fetch::Body { move(stream), move(source), move(length) };
+    // 10. Return (body, type).
+    return Fetch::BodyWithType { .body = move(body), .type = move(type) };
```

Comment:
nit: you don't need the explicit typename here
```suggestion
    return { .body = move(body), .type = move(type) };
```